src/devices/cpu/arcompact/arcompact.h

```diff
@@ -105,795 +40,751 @@ class arcompact_device : public cpu_device
+	uint32_t arcompact_auxreg100_TIMER1_r(offs_t offset);
+	void arcompact_auxreg012_TIMER0_w(offs_t offset, uint32_t data);
+	void arcompact_auxreg100_TIMER1_w(offs_t offset, uint32_t data);
+
+	uint32_t arcompact_auxreg025_INTVECTORBASE_r();
+	void arcompact_auxreg025_INTVECTORBASE_w(uint32_t data);
+
+	uint32_t arcompact_auxreg043_AUX_IRQ_LV12_r();
+	void arcompact_auxreg043_AUX_IRQ_LV12_w(uint32_t data);
+
+	uint32_t arcompact_auxreg200_AUX_IRQ_LVL_r();
+	void arcompact_auxreg200_AUX_IRQ_LVL_w(uint32_t data);
+
+
+	const static int REG_BLINK    = 0x1f; // r31
+	const static int REG_GP       = 0x1a; // r26
+	const static int REG_FP       = 0x1b; // r27
+	const static int REG_SP       = 0x1c; // r28
+	const static int REG_ILINK1   = 0x1d; // r29
+	const static int REG_ILINK2   = 0x1e; // r30
+	const static int REG_MLO      = 0x39; // r57 - multiply low 32-bits (of 64-bit result)
+	const static int REG_MMID     = 0x3a; // r58 - multiply mid 32-bits (of 64-bit result - overlaps)
+	const static int REG_MHI      = 0x3b; // r59 - multiply high 32-bits (of 64-bit result)
+	const static int REG_LP_COUNT = 0x3c; // r60
+	const static int REG_LIMM     = 0x3e; // r62 - used to indicate long immedaite
+	const static int REG_PCL      = 0x3f; // r63
+
+	const static uint32_t E1_FLAG = 0x00000002;
+	const static uint32_t E2_FLAG = 0x00000004;
+
+	const static uint32_t V_OVERFLOW_FLAG = 0x00000100;
+	const static uint32_t C_CARRY_FLAG = 0x00000200;
+	const static uint32_t N_NEGATIVE_FLAG = 0x00000400;
+	const static uint32_t Z_ZERO_FLAG = 0x00000800;
+
+	enum
+	{
+		ARCOMPACT_PC = STATE_GENPC,
+		ARCOMPACT_STATUS32 = 0x10,
+		ARCOMPACT_LP_START,
+		ARCOMPACT_LP_END
+	};
+
+	void get_limm_32bit_opcode(void)
+	{
+		m_regs[REG_LIMM] = (READ16((m_pc + 4)) << 16);
+		m_regs[REG_LIMM] |= READ16((m_pc + 6));
+	}
+
+	void get_limm_16bit_opcode(void)
+	{
+		m_regs[REG_LIMM] = (READ16((m_pc + 2)) << 16);
+		m_regs[REG_LIMM] |= READ16((m_pc + 4));
+	}
+
+	// registers used in 16-bit opcodes have a limited range
+	// and can only address registers r0-r3 and r12-r15
+	uint8_t common16_get_and_expand_breg(uint16_t op)
+	{
+		uint8_t reg = ((op & 0x0700) >> 8);
+		if (reg>3) reg += 8;
+		return reg;
+	}
+
+	uint8_t common16_get_and_expand_creg(uint16_t op)
+	{
+		uint8_t reg = ((op & 0x00e0) >> 5);
+		if (reg>3) reg += 8;
+		return reg;
+	}
+
+	uint8_t common16_get_and_expand_areg(uint16_t op)
+	{
+		uint8_t reg = ((op & 0x0007) >> 0);
+		if (reg>3) reg += 8;
+		return reg;
+	}
+
+	uint8_t common16_get_u3(uint16_t op)
+	{
+		return ((op & 0x0007) >> 0);
+	}
+
+	uint8_t common16_get_u5(uint16_t op)
+	{
+		return ((op & 0x001f) >> 0);
+	}
+
+	uint8_t common16_get_u8(uint16_t op)
+	{
+		return ((op & 0x00ff) >>0);
+	}
+
+	uint8_t common16_get_u7(uint16_t op)
+	{
+		return ((op & 0x007f) >>0);
+	}
+
+	uint8_t common16_get_s9(uint16_t op)
+	{
+		uint8_t s = ((op & 0x01ff) >> 0);
+		if (s & 0x100)
+			s |= 0xfe00;
+
+		return s;
+	}
```

Comment:
Please `util::sext` for sign extending arbitrary-size fields.  Also, I’m pretty sure the return type on this is wrong – you’re returning a value from a nine-bit fields in a `uint8_t`.

Reply:
fixed it